날씨 정보(전) [STEP 3] 산체스

sanches37 opened the pull request:
@jungseungyeo 

안녕하세요 린생~!
고민을 하다가 너무 늦어질 것 같아서, 부족하지만 PR을 보냅니다.

### 고민한점
뷰의 레이아웃이 변경되는 것에 대비하기 쉽고,뒷 스탭에서 가로화면 전환 구현이 있어서, TableView 보다 CollectionView 로 구현하면 가로화면에서도 레이아웃 구현하는데 좋다고 생각해서 CollectionView 로 구현하였습니다.

이미지캐시를 적용했는데, 현 상태에서는 이것만으로도 지연로딩에 문제가 되지 않을 거라 생각해서 따로 지연로딩에 관한 코드를 구현하지 않았습니다.

리프레쉬 컨트롤러를 구현했는데, 작동이 원할하지 않아서 고민을 오래하던 중 PR을 보내고 린생께 여쭤보는게 빠르겠다고 생각했습니다.
리프레쉬 컨트롤을 실행했을 때 경도, 위도 값을 아에 못받아오거나 일부만 받아오는 현상이 생겼습니다. 
제대로 못받아왔을 때 다시 리프레쉬 컨트롤을 실행하면 온전히 받아오게 됐는데, 이 부분에서 정확한 문제점을 파악하지 못했습니다.
여러군데서 받아오는 데이터들의 시점이 문제가 될까봐, 디스패치 그룹을 사용해서 모든게 받아온 뒤 화면을 리프레쉬 할 수 있도록 코드를 수정했지만, 문제가 해결 되지 않았습니다. 그러던 중 리프레쉬 컨트롤을 작동했을 때 디스패치 그룹 내부에서 실행하는 코드들이 두번씩 엇갈려 호출되서 문제가 있다는걸 알게됐습니다.

```swift
private func processWeatherImpormation(completion: @escaping () -> Void) {
       
            guard let location = currentLocation else { return }
            let currentWeatherURL =
            WeatherURL.weatherCoordinates(latitude: location.coordinate.latitude,
                                          longitude: location.coordinate.longitude)
            let fiveDaysWeatherURL =
            WeatherURL.forecastCoordinates(latitude: location.coordinate.latitude,
                                           longitude: location.coordinate.longitude)
        
        DispatchQueue.global().async {
            let group = DispatchGroup()
            group.enter()
            self.getWeatherImpormation(request: fiveDaysWeatherURL,
                                  type: FiveDaysWeather.self) { result in
               
                if let result = result {
                    self.collectionViewDataSource.fiveDaysWeather = result
                }
                print("1")
                group.leave()
            }
            
            group.enter()
            self.getWeatherImpormation(request: currentWeatherURL,
                                  type: CurrentWeather.self) { result in
                
                if let result = result {
                    self.collectionViewDataSource.currentWeather = result
                }
                print("2")
                group.leave()
            }
            
            group.enter()
            self.locationManager.getUserAddress(location: location) { address in
                switch address {
                case .success(let data):
                    self.collectionViewDataSource.currentAddress = data
                case .failure(let error):
                    self.handlerError(error)
                }
                print("3")
                group.leave()
            }
            
            group.wait()
            print("4")
            completion()
        }
```
위 코드를 실행했을 때 앱 로딩 후에는 정상적으로 프린트 출력이 되었지만, 이후 리플레쉬 컨트롤 작동 시 엇갈려 호출되는걸 알게됐습니다.
현재 코드상에서 여러번 호출되는 상황이 무엇 때문인지 고민을 하다가 리플레시 컨트롤을 정상적으로 구현하지 못한채 PR을 보내게 됐습니다.

앱 실행시 print 출력

<img width="616" alt="스크린샷 2021-10-19 오후 7 15 42" src="https://user-images.githubusercontent.com/84059338/137890530-af8f3b0b-56c6-4a0a-ac33-33a2d2aeadb3.png">

리플레시 컨트롤 작동시 출력

<img width="654" alt="스크린샷 2021-10-19 오후 7 17 52" src="https://user-images.githubusercontent.com/84059338/137890791-b95f25f1-fa17-4545-8164-f75657d2bb94.png">

이번 리뷰도 잘 부탁드립니다 😊🙏🏼

sanches37 committed:
feat: 현재 날씨와 5일 예보의 타입 구현
CurrentWeather, FiveDaysWeather 타입 구현

sanches37 committed:
refactor: Models 디렉토리 생성 후 이동
CommonWeatherModels.swift 생성 후 각 타입별 세부타입 분리

YongHoonJJo committed:
feat: JSONDecodable 프로토콜 추가

- readLocalFile(for:) 메서드 추가 : 로컬 파일을 읽어서 Data 형태로 만들기 위한 메서드
- decodeJSON(_:from:) - JSON 디코딩을 위한 메서드
- ParsingError 열거형 타입 추가 : 위 메서드에서 발생하는 에러 적용.

sanches37 committed:
feat: 범용성을 고려한 네트워킹 타입 구현
- APIGetCallable 프로토콜 구현 및 extension 을 통한 callGetAPI 메서드 추가

YongHoonJJo committed:
feat: APIError 열거형 타입 구현

- LocalizedError 프로토콜 채택 및 errorDescription 연산프로퍼티 적용.
- callGetAPI(_:_:url:completion:) 으로 레이블 수정
- callGetAPI(_:_:url:completion:) 메서드 내 .failure(Error) 적용

sanches37 committed:
refactor: 날씨 모델에 대한 누락된 프로퍼티 추가

sanches37 committed:
feat: 에러타입 추가 및 기존모델 프로퍼티 옵셔널로 변경
- WeatherError 타입 추가

YongHoonJJo committed:
refactor: JSONDecodable 타입 구조 변경

- 기존 JSONDecodable 의 protocol + extension 조합 해제.
- JSONDecodable 프로토콜을 채택한 DecodingManager 타입으로 변경.
- 불필요한 메서드 삭제 : readLocalFile(for:) throws -> Data
- 파일 이름 변경 : JSONDecodable.swift -> DecodingManager.swift
- APIGetCallable.swift 에서 callGetAPI() 메서드 내의 decodeManager.decodeJSON() 메서드 활용 방식 변경
  > 의존성 주입을 통한 메서드 호출방식으로 변경.

YongHoonJJo committed:
refactor: APIResource 타입 추가 및 APIManager 에서 API 호출 메서드 구조 및 네이밍 수정.

- HTTP Method 에 대한 열거형 타입 추가.
- url, method 를 받을 수 있는 APIResource 타입 추가.
- APIGetCallable protocol+extension 조합을 매니저 클래스로 수정.
  > callGetAPI() 메서드 네이밍 수정. requestAPI() 로 수정.
- requestAPI() 메서드 매개변수 변경
  > url 삭제, APIResource 타입 매개변수 추가.
  > 내부에서 request 생성하는 코드 추가.

sanches37 committed:
# <타입>: <제목>
feat: 날씨 URL 생성타입 및  ErrorMessageProtocol 파일추가
##### 제목은 최대 50 글자까지만 입력 ############## -> |
- enum WeatherConfig, enum WeatherURL 타입 추가
- protocol ErrorMessageProtocol 타입 추가

# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# 꼬릿말은 아래에 작성: ex) #이슈 번호

# --- COMMIT END ---
# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# ------------------
#     제목 첫 글자를 대문자로
#     제목은 명령문으로
#     제목 끝에 마침표(.) 금지
#     제목과 본문을 한 줄 띄워 분리하기
#     본문은 "어떻게" 보다 "무엇을", "왜"를 설명한다.
#     본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
# ------------------

sanches37 committed:
# <타입>: <제목>
refactor: weatherError 메시지 프로토콜 채택 및 initialize 추가
##### 제목은 최대 50 글자까지만 입력 ############## -> |
- struct WeatherError 타입에 ErrorMessageProtocol 타입 채택 및 준수
- ErrorMessageProtocol 프로토콜을 APIManager 에서 활용 및 일부 로직 수정
- weatherError 타입에 Decodable 의 initializer 구현

# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# 꼬릿말은 아래에 작성: ex) #이슈 번호

# --- COMMIT END ---
# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# ------------------
#     제목 첫 글자를 대문자로
#     제목은 명령문으로
#     제목 끝에 마침표(.) 금지
#     제목과 본문을 한 줄 띄워 분리하기
#     본문은 "어떻게" 보다 "무엇을", "왜"를 설명한다.
#     본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
# ------------------

sanches37 committed:
# <타입>: <제목>
feat: WeatherURL 타입에 대응하는 APIResource init 추가
##### 제목은 최대 50 글자까지만 입력 ############## -> |

# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# 꼬릿말은 아래에 작성: ex) #이슈 번호

# --- COMMIT END ---
# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# ------------------
#     제목 첫 글자를 대문자로
#     제목은 명령문으로
#     제목 끝에 마침표(.) 금지
#     제목과 본문을 한 줄 띄워 분리하기
#     본문은 "어떻게" 보다 "무엇을", "왜"를 설명한다.
#     본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
# ------------------

sanches37 committed:
# <타입>: <제목>
refactor: 타입 정정
##### 제목은 최대 50 글자까지만 입력 ############## -> |
- CommonWeatherModels.swift > List 타입 내  pop 프로퍼티에 대한 타입 변경

# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# 꼬릿말은 아래에 작성: ex) #이슈 번호

# --- COMMIT END ---
# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# ------------------
#     제목 첫 글자를 대문자로
#     제목은 명령문으로
#     제목 끝에 마침표(.) 금지
#     제목과 본문을 한 줄 띄워 분리하기
#     본문은 "어떻게" 보다 "무엇을", "왜"를 설명한다.
#     본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
# ------------------

sanches37 committed:
# <타입>: <제목>
etc: 리뷰를 위한 테스트코드 적용
##### 제목은 최대 50 글자까지만 입력 ############## -> |

# 본문은 위에 작성
######## 본문은 한 줄에 최대 72 글자까지만 입력 ########################### -> |

# 꼬릿말은 아래에 작성: ex) #이슈 번호

# --- COMMIT END ---
# <타입> 리스트
#   feat    : 기능 (새로운 기능)
#   fix     : 버그 (버그 수정)
#   refactor: 리팩토링
#   style   : 스타일 (코드 형식, 세미콜론 추가: 비즈니스 로직에 변경 없음)
#   docs    : 문서 (문서 추가, 수정, 삭제)
#   test    : 테스트 (테스트 코드 추가, 수정, 삭제: 비즈니스 로직에 변경 없음)
#   chore   : 기타 변경사항 (빌드 스크립트 수정 등)
# ------------------
#     제목 첫 글자를 대문자로
#     제목은 명령문으로
#     제목 끝에 마침표(.) 금지
#     제목과 본문을 한 줄 띄워 분리하기
#     본문은 "어떻게" 보다 "무엇을", "왜"를 설명한다.
#     본문에 여러줄의 메시지를 작성할 땐 "-"로 구분
# ------------------

YongHoonJJo committed:
refactor: UrlPathGeneratable 프로토콜 선언

- WeatherURL 타입에 UrlPathGeneratable 채택 후 준수.
- APIResource 의 이니셜라이저에 WeatherURL 타입대신 추상타입인 UrlPathGeneratable 로 적용.

YongHoonJJo committed:
refactor: APIResource 타입에 접근제어 추가.

YongHoonJJo committed:
refactor: RequestGeneratable 프로토콜 추가

- APIResource 타입에 RequestGeneratable 프로토콜 채택
- APIManager 에서 매개변수 APIResource 타입을 RequestGeneratable 타입으로 변경하여 추상화 적용.

YongHoonJJo committed:
refactor: 프로토콜 타입 별도의 파일로 분리.

sanches37 committed:
refactor: 의존성주입을 통해 테스트가 가능한 API 객체로 변경

- URLSessionProtocol 프로토콜 추가
- URLSession 에 URLSessionProtocol 프로토콜 채택
-  APIManager 에서 URLSessionProtocol 타입을 의존성 주입할 수 있도록 코드 수정

YongHoonJJo committed:
refactor: Error Handle 로직 이동.

- APIError 타입에 커스텀 이니셜라이저를 통한 처리.
- APIError 타입 별도의 파일로 분리.

YongHoonJJo committed:
refactor: 불필요한 주석 코드 삭제.

YongHoonJJo committed:
feat: 에러 객체를 위한 ErrorHandle 타입 추가.

- UIViewController+HandleError extension 추가.

sanches37 committed:
etc: podfile 을 위한 gitignore 적용

sanches37 committed:
etc: swiftLint 설치 및 적용

sanches37 committed:
etc: swiftlint.yml 파일 생성 및 disabled rules 적용

YongHoonJJo committed:
etc: cocoapod update 후 pod install 재 실행.

YongHoonJJo committed:
refactor: dataTask() 의 클로즈 내부 로직중 강제언래핑 로직 변경.

sanches37 committed:
refactor: URLComponent 를 활용한 URL 생성 로직 수정

- generateURLPath -> generateURL 로 네이밍 변경
- generateURL 내 URLComponent 로직으로 수정 및 리턴 타입 URL? 로 변경
- APIResource 내  url 프로퍼티 타입 URL? 로 변경
- UrlPathGeneratable 프로토콜 -> UrlGeneratable 로 네이밍 변경 및 함수 프로토 타입 네임 및 리턴 타입 수정

sanches37 committed:
Merge commit '9c3c39c3607e7d370ac07939cf92f90532bedda9' into Step-1

sanches37 committed:
refactor: dataTask 의 closure 내 error 매개변수에 대한 에러 처리 case 변경

- APIError 내 사용하지 않는 dataTask case 삭제
- APIError의 unknown case 에 String 연관값으로 변경

sanches37 committed:
refactor: URLRequest 생성 로직 수정

- UrlGeneratable 프로토콜 프로퍼티 추가
-  APIResource initializer 파라미터 수정, generateRequest 메서드 로직 수정
-  WeatherURL 타입 내 parameter 로직 구현, generateURL 메서드 로직 수정

sanches37 committed:
feat: LocationManager 클래스 추가

- 현재위치 획득을 위한 LocationManager 추가

sanches37 committed:
refactor: WeatherURL enum 수정

- case 가 상단에 오도록 순서 변경

sanches37 committed:
refactor: APIResource 타입 코드 수정

- components 의 항목을 String으로 변환해서 받았던 구조를 url 그대로 받을 수 있도록 수정
- queryString 이 배열에 담기지 않아서 코드 추가

sanches37 committed:
refactor: LocationManager 권한요청 방식 변경

- locationManager(didChangeAuthorization status:) 메서드를 이용해서 권한 요청하도록 변경
- viewController -> WeatherImpormationViewController 로 이름 변경
- info.plist 권한요청 추가

sanches37 committed:
feat: LocationManager 주소 전달 메서드 생성

- getUserAddress 생성
- LocationManager 위치 정리
- LocationError enum 생성

sanches37 committed:
refactor: getUserAddress 메서드 String 을 리턴하도록 수정

- 필요한 정보인 administrativeArea, thororoughfare 만 String 으로 전달할 수 있도록 구조 변경
- LocationError 타입 내 invalidAddress case 추가

sanches37 committed:
refactor: 정보 업데이트 방식 변경

-  앱의 특성에 따라 한번 호출에 적합한 requestLocation 메서드로 변경
- 정보 업데이트 코드를 getUserLocation 에서 사용했는데, didChangeAuthorization 에서 사용하도록 변경
- didFailWithError 를 추가하여 위치가 받아지지 않았을 때 error 처리 코드 추가

sanches37 committed:
refactor: LocationManager 구조변경

- locationManager.delegate = self 코드를 init 에서 하도록 수정
- 추후 화면 refresh 를 위한 getUserLocation() 에 requestLocation() 추가
- 위치정보를 받지 못했을 때 재시도를 위해서 didfailWithError 에 requestLocation() 추가

sanches37 committed:
refactor:  APiError 에 Error 프로토콜 채택이 안된 것 추가

sanches37 committed:
refactor: 에러핸들러 디버그프린트로 변경

- 기존 print 대신 debugPrint 를 사용하도록 수정

sanches37 committed:
feat: viewController 에서 날씨정보를 가져오는 코드 추가

- 주소를 가져오는 getUserAddress() 추가
- 날씨정보를 네트워크로 가져오는 getCurrentWeatherImpormation() 추가
- 날씨정보를 네트워크로 가져오는 getFiveDaysWeatherImpormation() 추가
- 날씨정보 및 주소를 한번에 최신화 할 수 있는 processWeatherImpormation() 추가

sanches37 committed:
refactor: LocationManager 테스트 위한 코드 추가

- Unit Test 코드 구현
- MockLocationManager 구현
- 의존성 주입을 위한 LocationManagerProtocol, LocationManagerDelegate 프로토콜 구현
- 의존성 주입을 위해 LocationManager 타입 수정

sanches37 committed:
refactor: API 호출 함수 하나로 통합

- getFiveDaysWeatherImpormation, ,getCurrentWeatherImpormation -> getWeatherImpormation 로 메서드 통합

sanches37 committed:
refactor: processWeatherImpormation() 함수 내부 분리

- WeatherImpormationViewController 내 processWeatherImpormation() 메서드가 많은 작업을 하고 있어서 코드 분리
- API 호출을 통한 코드는 processWeatherImpormation() 메서드에서 처리하도록 수정
- locationManager.getUserLocation 의 호출은 getUserLocation 에서 처리하도록 수정
- processWeatherImpormation, getUserAddress 는 ViewDidAppear 에서 처리하도록 수정
- locationManager.getUserAddress 메서드 매개변수 받는 구조로 수정

sanches37 committed:
refactor: getUserLocation 메서드 weak self 제거

sanches37 committed:
refactor: didFailWithError 내 requestLocation 제거

sanches37 committed:
refactor: getUserLocation() 의 클로저에서 함수 호출하도록 변경

- 작동 불가로 인해 getUserLocation() 의 클로저에서 함수 호출하도록 변경

sanches37 committed:
refactor: 날씨정보 및 주소 프로퍼티 옵저버에서 호출하도록 변경

- currentLocation 의 didSet 에서 processWeatherImpormation(), getUserAddress() 호출하도록 변경

sanches37 committed:
test: Test 코드 수정

-  클로져 안에서 결과 값을 도출하는 것을 외부에서 결과를 처리하는 방식으로 수정

sanches37 committed:
feat: CollectionView 기본구조 추가

- FiveDaysWeatherCell 추가
- FiveDaysWeatherCell 에서 사용할 작업할 부분 extension (DateFormatter, MeasurementFormatter, CALayer)
- WeatherImpormationViewController 내 collectionView 레이아웃 지정 및 register 등록
- WeatherCollectionViewDataSource 타입 생성

sanches37 committed:
feat: CollecionViewDataSource 구현 및 CompositionalLayout 적용

- CompositionalLayout 타입 생성 후 ViewController 내부에서 함수 호출

sanches37 committed:
refactor: URLSession dataTask 호출 구조

- DataTask 에러처리를  extension URLSessionProtocol 내부 함수로 호출하도록 수정
- APIManager Struct 로 변경

sanches37 committed:
feat: ImageManager 타입 구현

sanches37 committed:
feat: 헤더뷰 타입 구현 및 WeatherCollectionViewDataSource 내 코드추가

sanches37 committed:
refactor: CollectionViewCell Image 추가

-imageManager 를 cell 내에서 호출한 후 weatherImage 에 넣어줌

sanches37 committed:
feat: 헤더뷰 항목들 화면에 표시되도록 코드 작성

- locationManager 에서 전달하는 주소를 currentAddress 프로퍼티에 저장 후 headerView 의 addressLabel 에 전달
- headerView 오토레이 아웃 수정
- cell 과 같은 방식으로 label 과 image 적용

sanches37 committed:
feat: refreshController 구현

- setUpRefreshControl 및 handleRefreshControl 메서드 구현
- 한번에 reload 하기 위해 디스패치 그룹추가
- 백그라운드 화면 추가
- 백그라운드 화면으로 인해 label 색깔 화이트로 지정

sanches37 committed:
refactor: CollectionViewCell Contents nil 작업 추가

jungseungyeo commented on WeatherForecast/WeatherForecast/Views/FiveDaysWeatherCell.swift:
```diff
@@ -0,0 +1,128 @@
+        
+        private func imageConfigure(list: List?) {
+            guard let icon = list?.weather?.first?.icon else { return }
+            imageManager.fetchImage(url: ImageURL.weather(icon).path) { image in
```
```suggestion
            imageManager.fetchImage(url: ImageURL.weather(icon).path) { result in
```
로 하는게 좀 더 명확한 네이밍인거같아요.

sanches37 replied:
result 가 더 명확한 네이밍인 것 같습니다. 수정했습니다. 😀

jungseungyeo commented on WeatherForecast/WeatherForecast/Managers/ImageManager.swift:
```diff
@@ -0,0 +1,49 @@
+        self.session = URLSession(configuration: configuration)
+    }
+    
+    func fetchImage(
```
cancel은 따로 구현하지 않느 건가요?

sanches37 replied:
requestCachePolicy = .returnCacheDataElseLoad 를 사용하여 캐시를 우선사용하도록 정책을 사용해서 추가 코드 없이 이미지 캐시가 이루어진다고 생각되었는데요.
```swift
let request = URLRequest(url: url, cachePolicy: .returnCacheDataElseLoad)
        
        if let response = cache.cachedResponse(for: request),
           let imageData = UIImage(data: response.data) {
            compleHandler(.success(imageData))
            return nil
        } else {
            let dataTask = session.dataTask(with: request) { data, response, error in
                if let convertResponse = response, let convertData = data {
                    self.cache.storeCachedResponse(
                        CachedURLResponse(
                            response: convertResponse, data: convertData), for: request)
                }
```
위와 같이 이미지가 있으면 캐시의 이미지를 쓰고 아니면 dataTask 로 받아오는 코드로 작성하라는 말씀이실까요?

dataTask를 cancel 하는 코드가 없다는 의미라면, URLSessionDataTask 를 반환값으로 가져서 cell에서 사용시 cancel 하는 코드를 추가할 생각입니다. 🙂

jungseungyeo commented on WeatherForecast/WeatherForecast/CollectionView/WeatherCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,63 @@
+            for: indexPath) as? FiveDaysWeatherCell else {
+                return UICollectionViewCell()
+            }
+        cell.resetContents()
```
여기서 resetContents()를 호출하는 것보다 cell 내부에서 해주는 것이 좋아요. controller에서 작성하게 되면 controller가 cell의 대해서 좀 더 알아야되어 의존성이 커지게 되는거 같아요.

sanches37 replied:
resetContents 대신 prepareForReuse 로 셀이 재사용되기 전에 초기화하도록 수정하였습니다. 괜찮을까요? 🙂

jungseungyeo commented on WeatherForecast/WeatherForecast/CollectionView/WeatherCollectionViewDataSource.swift:
```diff
@@ -0,0 +1,63 @@
+
+import UIKit
+
+class WeatherCollectionViewDataSource: NSObject {
```
따로 DataSource를 만든 이유가 있을까요?

sanches37 replied:
뷰컨트롤러가 많은 일을 처리하게 될 것 같아서 DataSource 나 Delegate 를 분리하여 사용하려는 의도였습니다. 😊

jungseungyeo commented on WeatherForecast/WeatherForecast/Controllers/WeatherImpomationViewController.swift:
```diff
@@ -0,0 +1,169 @@
+    private let collectionViewDataSource = WeatherCollectionViewDataSource()
+    private var currentLocation: CLLocation? = nil {
+        didSet {
+            processWeatherImpormation {
```
fetch라는 함수를 만들어서 여기서는 호출만 해주면 좀 더 코드가 깔끔해질거 같아요.

sanches37 replied:
didset 내부에서 fetchWeatherImpormation 메서드만 호출하도록 변경했습니다. 😃

jungseungyeo commented on WeatherForecast/WeatherForecast/Controllers/WeatherImpomationViewController.swift:
```diff
@@ -0,0 +1,169 @@
+            WeatherURL.forecastCoordinates(latitude: location.coordinate.latitude,
+                                           longitude: location.coordinate.longitude)
+        
+        DispatchQueue.global().async {
```
DispatchQueue.global().async를 사용하는 이유가 있나요?

sanches37 replied:
DispatchQueue.global().async 내부 코드들이 비동기적으로 호출되도록 하기 위해서 사용했습니다. 😀

jungseungyeo commented on WeatherForecast/WeatherForecast/Controllers/WeatherImpomationViewController.swift:
```diff
@@ -0,0 +1,169 @@
+                group.leave()
+            }
+            
+            group.wait()
```
wait()를 사용하지 않고 notifyf를 사용하는 방법은 어떤가요?

sanches37 replied:
notify 로 그룹 완료 후 비동기적으로 사용하면서 DispatchQueue.global().async 코드를 없앨 수 있었습니다. 😆

jungseungyeo commented on WeatherForecast/WeatherForecast/Extensions/DateFormatter + convertDate.swift:
```diff
@@ -0,0 +1,17 @@
+    func convertDate(intDate: Int) -> String {
+        let date = Date(timeIntervalSince1970: TimeInterval(intDate))
+        locale = Locale(identifier: "Ko-kr")
+        dateFormat = "MM/dd(E) HH시"
```
이런 형식을 무엇이라고 하나요? 따른 형식은 어떤게 있을까요? 
https://ko.wikipedia.org/wiki/ISO_8601 국제 표준 날짜 표시내용이 있어요 한번 참고해주세요 :)

sanches37 replied:
공유해주신 내용을 읽어보니 날짜와 시간과 관련된 데이터 교환을 다루는 국제 표준이군요. 감사합니다.
ISO8601DateFormatter.formatOptions = [.withDay, .withMonth, .withYear, .withTime] 로 지정한 요소만 보일 수 있게 하는 방법도 있더군요. ISO8601DateFormatter 도 사용해보려 했지만, 요일이 없는 것 같아서 현재구조를 유지하는 방향으로 정했습니다. 😃

jungseungyeo commented on WeatherForecast/WeatherForecast/Layout/CompositionalLayout.swift:
```diff
@@ -0,0 +1,64 @@
+}
+
+struct CompositionalLayout {
+    func create(
```
파라미터 갯수가 많고 함수의 크기가 너무 큰거 같아요. 한 번 줄여서 사용하는 방법을 사용해주세요

sanches37 replied:
파라미터의 갯수를 줄이기 위해 protocol, enum 을 이용해 전달할 타입을 만들고, 매개변수로 전달하는 구조로 변경하였습니다.
메서드 크기를 줄이기 위해 내부 item, group, header 에 해당하는 부분을 각각 메서드화 해서 create 내에서 호출하도록 하였습니다.
위와같은 방식이 유도하셨던 방법이 맞는지 모르겠습니다. 😅

jungseungyeo commented on WeatherForecast/WeatherForecast/Managers/LocationManager.swift:
```diff
@@ -0,0 +1,81 @@
+}
+
+extension LocationManager: LocationManagerDelegate {
+    func locationManager(_ manager: CLLocationManager,
```
여기서 위치 정보가 리플레쉬또는 위치 정보 요청시 여러번 호출이 되고 있어요. 흠.. 실제로는 한번만 호출되어야 하는데.. 이상하네요 ㅎㅎ..

sanches37 replied:
공유해주신 자료로 debounce 를 이해하고, 같은 문제를 가지고 있던 조스트의 코드를 참고하여 debounce 를 구현해서 리플레시 컨트롤이 잘 작동하게 수정하였습니다. 감사합니다. 😆

jungseungyeo commented on WeatherForecast/WeatherForecast/Views/CurrentWeatherHeaderView.swift:
```diff
@@ -0,0 +1,144 @@
+    
+    private func imageConfigure(weather: CurrentWeather?) {
+        guard let icon = weather?.weather?.first?.icon else { return }
+        imageManager.fetchImage(url: ImageURL.weather(icon).path) { image in
```
cell이 재사용 할 때 마다 ulr을 요청하네요. cancel하고 cache 로직을 추가하면 좋을거같아요.

sanches37 replied:
기존 코드에서도 메모리캐시가 되고 있다고 생각했는데, 코드상 확실히 캐시처리를 하고 있는지 알기 위해 imageManager 의 구조를 수정하고 확인해보니 캐시처리가 되고 있는 걸 확인했습니다. imageConfigure 메서드에서 해당 주소를 넣어주면 fecthImage 내부에서 판별 후 캐시된 이미지를 쓸지 네트워크로 받을지를 판별하는데 이런 구조가 괜찮을까요?
또 지연로딩 처리를 위해 fetchImage 리턴값을 가지게 구조를 바꿨습니다. prepareForReuse에서 이미지가 셀에 적용하기 전에, 셀이 재사용되거나 하는 경우 가지고 온 이미지를 무시하도록 코드를 적용했습니다.😀

jungseungyeo commented on WeatherForecast/WeatherForecast/Views/FiveDaysWeatherCell.swift:
```diff
@@ -0,0 +1,128 @@
+    var path: String {
+        switch self {
+        case .weather(let id):
+            return "https://openweathermap.org/img/w/\(id).png"
```
이곳보다는 모델에서 저장 할때 가지고 있는게 좋을거같아요.

sanches37 replied:
View인 cell 에서 가지고 있는게 맞지 않는 것 같습니다. imageManager 폴더 내로 이동하였습니다. 😊

jungseungyeo commented on WeatherForecast/WeatherForecast/Layout/CompositionalLayout.swift:
```diff
@@ -0,0 +1,64 @@
+        }
+        
+        let layout =
+            UICollectionViewCompositionalLayout { (_, _) -> NSCollectionLayoutSection? in
```
UICollectionViewCompositionalLayout를 사용한 이유가 있나요?
이번 학습 목표는 아니여서 궁금합니다.

sanches37 replied:
따로 특별한 이유는 없고, 여러 레이아웃에 직관적으로 사용하기 편했고, 컬렉션뷰의 헤드뷰나 섹션이 나눠줬을 때도 쉽게 대응할 수 있어서 컴포지셔널 레이아웃을 선택해서 구현해봤습니다. 다만 IOS 13 이후 사용이라는 점이 좀 걸렸는데, IOS 14 이전 기능은 그냥 사용하자 싶어서 적용해보았습니다. 😃

jungseungyeo commented on WeatherForecast/WeatherForecast/Protocols/LocationProtocol.swift:
```diff
@@ -0,0 +1,35 @@
+}
+
+extension CLLocationManager: LocationManagerProtocol {
+    var locationManagerDelegate: LocationManagerDelegate? {
```
step2에서 이 delegate가 weak가 아니였어요. weak를 붙여주면 좋을거같아요. 또 왜 weak를 붙어주어야할까요?

sanches37 replied:
변경했습니다! weak 이 없으면  self.locationManager.locationManagerDelegate = self 코드에서 locationManagerDelegate 와 LocationManager 가 순환 참조되어 메모리 해제가 안될 것 같습니다. 😃

jungseungyeo reviewed (changes requested):
산체스 step3고생하셨습니다. 

step3를 하면서 많은 고민을 하고 구현을 하였다고 느꼈습니다. :) 동시성으로 api를 3개를 기다리는 것을 구현하는 방식을 사용한 것도 좋았습니다.

고민한점

> 고민한 부분을 실제로 보았는데요. 시뮬레이터에서만 나오고 실제 디바이스 빌드시 발생을 하지 않네요.. 정확한 원인은 모르겠지만 해결을 한 번 해보았으면 좋겠습니다.  

https://webclub.tistory.com/607

Debounce와 Throttle라는 개념이 있습니다. 이것을 직접 구현하여 해결해보면 좋을거 같아요. :) 

또 코멘트에서 이야기 한 거 처럼 이미지 캐시와 imagerequest cancel을 구현을 해주세요. 이번 학습에서 꼭 필요한 부분이라 요청드립니다.  또 코멘트에 답변을 달아주세요.

sanches37 committed:
refactor: imageConfigure 내 네이밍 변경

- completion 으로 전달하는 값의 네이밍을 image -> result 로 변경

sanches37 committed:
refactor: resetContents 대신 prepareForReuse 사용

- label 과 image 초기화를 위해 prepareForReuse 사용하도록 수정

sanches37 committed:
refactor: didset 내부 호출함수 구조 변경

- fetchWeatherImpormation 함수만 didset 내부에서 호출하도록 구조 변경

sanches37 committed:
refactor: DispatchGroup 구조 수정

- wait 대신 notify 를 사용하게 구조 수정

sanches37 committed:
refactor: enum ImageURL 을 imageManager.swift 로 이동

sanches37 committed:
refactor: ImageManager 지연로딩 및 캐시 위해 코드 수정

- 이미지 캐시처리 및 확인 위해 코드 수정
- 지연로딩 위한 fetchImage 리턴값을 가지게 수정 후 cell 에서 cancel 할 수 있도록 수정

sanches37 committed:
refactor:  var locationManagerDelegate 에 약한 참조 추가

sanches37 committed:
fefactor: debounce 구현

- debounce 구현하여 requestLocation 여러번 호출되는 거 방지하여 리플레시 컨트롤 정상 작동되게 수정

sanches37 committed:
refactor: @discardableResult 추가

- ImageManager 타입의 fetchImage 메서드의 리턴값이 사용안되는 경우가 있어서 @discardableResult 추가

sanches37 committed:
refactor: CompositionalLayout 타입 구조 수정

- 매개변수를 줄이기 위해 CompositionalLayoutContents 에 해달하는 enum 생성
- 매개변수를 줄이기 위해 WeatherImpormationLayout 생성 후 CompositionalLayoutProtocol 채택
- create 메서드 크기를 줄이기 위해 메서드 분리

jungseungyeo commented on WeatherForecast/WeatherForecast/Controllers/WeatherImpomationViewController.swift:
```diff
@@ -72,10 +73,16 @@ class WeatherImpormationViewController: UIViewController {
+        if timer != nil {
+            timer?.invalidate()
+        }
+        timer = Timer.scheduledTimer(withTimeInterval: 0.3, repeats: false) { timer in
```
debounce는 여기가 아니고 delegate에서 호출해주면 좋을거같아요. 또 DispatchQueue와 DispatchWorkItem를 활요하면 좋을거같아요.

jungseungyeo replied:
https://stackoverflow.com/questions/27116684/how-can-i-debounce-a-method-call <- 여기를 참고하면 좋을거같아요.

jungseungyeo commented on WeatherForecast/WeatherForecast/Protocols/LocationProtocol.swift:
```diff
@@ -22,7 +22,7 @@ protocol LocationManagerDelegate: CLLocationManagerDelegate {
 
 extension CLLocationManager: LocationManagerProtocol {
-    var locationManagerDelegate: LocationManagerDelegate? {
+    weak var locationManagerDelegate: LocationManagerDelegate? {
```
너무 좋아요 :)

jungseungyeo commented on WeatherForecast/WeatherForecast/Controllers/WeatherImpomationViewController.swift:
```diff
@@ -123,9 +122,9 @@ class WeatherImpormationViewController: UIViewController {
-            group.wait()
-            completion()
-        }
+            group.notify(queue: .global()) {
```
너무 좋습니다 :)

sanches37 committed:
refactor: Compositional 레이아웃 구조 변경

- protocol CompositionalLayoutProtocol 수정
- CompositionalLayoutContents.swift 삭제
- WeatherImpormationLayout 삭제
- WeatherCollectionViewDataSource 내 레이아웃 적용 코드 수정

sanches37 committed:
refactor: Debounce 구조 수정

- dispatchWorkItem 을 사용하는 방식으로 변경
- DebounceManager 사용
- LocationManager 에서 debounce 적용

jungseungyeo commented on WeatherForecast/WeatherForecast/Controllers/WeatherImpomationViewController.swift:
```diff
@@ -69,7 +69,8 @@ class WeatherImpormationViewController: UIViewController {
             equalTo: view.safeAreaLayoutGuide.bottomAnchor).isActive = true
         
-        collectionViewDataSource.decidedLayout(collectionView)
+        collectionView.collectionViewLayout = layout.layout()
```
layout.layout이라는 네이밍은 이상하네요 ㅠ

jungseungyeo commented on WeatherForecast/WeatherForecast/Protocols/CompositionalLayoutProtocol.swift:
```diff
@@ -5,14 +5,8 @@
-    var scrollDirection: ScrollDirection { get }
-    var cellMargin: Margin? { get }
-    var viewMargin: Margin? { get }
+    func layout() -> UICollectionViewLayout
```
파라미터가 다 사라졌네요 ㅎㅎ 너무 좋아요. 이야기 한 거 처럼 구현해야되고 알아야되는게 많을 수록 의존성은 높아지고 복잡도가 높아지니 조심해주세요 :)

jungseungyeo commented on WeatherForecast/WeatherForecast/Managers/DebounceManager.swift:
```diff
@@ -0,0 +1,19 @@
+    private var queue: DispatchQueue = DispatchQueue(label: "Serial")
+    private var workItem = DispatchWorkItem(block: {})
+    
+    func processDebounce(interval: TimeInterval = 0.3, action: @escaping (() -> Void)) {
```
queue도 받아서 처리하면 더 좋았을거같네요 ㅎㅎ 그래도 너무 좋습니다.

jungseungyeo reviewed (approved):
산체스! 너무 고생하셨어요! 이번에 작성한 코드를 보면서 많은 노력을 보았습니다. protocol을 활요하고 역할을 나누는 부분에서 매우 잘 하셨습니다. 또한 bounce를 구현하여 좀 더 안정성이 있는 코드를 작성한 부분도 좋았습니다. 고생하셨습니다. 

protocol을 사용하는 것은 좋지만 구현해야되는 부분이 많아 지면 코드의 복잡도와 의존성이 높아 진다는 걸 잊지마세요! 수고하셨습니다.
